docs/CONFIG.md

### API_KEY


API key

Default unset, allowing unauthenticated access to APIs


Default: ``


### AWS_BUCKET_NAME

Name of s3 bucket for storing artifacts

Default: ``


### AWS_ENDPOINT

HTTPS url for AWS ENDPOINT for accessing S3. Set to minio API URL, if in use.

Default: ``


### BROKER_HOST

Broker hostname/ip

Default: ``


### BROKER_INSTANCE

Broker vhost

Default: ``


### BROKER_PASSWORD

Broker password

Default: ``


### BROKER_PORT

Broker port

Default: `0`


### BROKER_TYPE

Configure type of broker. Currently only supports 'amqp'

Default: `amqp`


### BROKER_USERNAME

Broker username

Default: ``


### CACHE_BROWSER


Whether to cache browser between runs.

This is experimental and might be unstable.

Use on non-production environment


Default: `False`


### CHECK_CRITICAL_THRESHOLD

Threshold for warning->critical check status in UI

Default: `99.0`


### CHECK_WARNING_THRESHOLD

Threshold for okay->warning check status in UI

Default: `99.9`


### CHROME_HEADLESS_MODE

Chrome headless mode. Options: None, New or Legacy

Default: `Legacy`


### DATABASE_HOST

Database hostname/IP

Default: ``


### DATABASE_NAME

Database DB name

Default: ``


### DATABASE_PASSWORD

Database password

Default: ``


### DATABASE_PORT

Database port

Default: ``


### DATABASE_TYPE

Database type - currently only 'postgresql' is supported

Default: ``


### DATABASE_USERNAME

Database username

Default: ``


### DEFAULT_CHECK_INTERVAL

Default check run interval, if not specified by check (seconds)

Default: `300`


### DEFAULT_CHECK_TIMEOUT

Default run timeout

Default: `60`


### FIREFOX_HEADLESS

Whether to run firefox in headless mode

Default: `True`


### MAX_CHECK_INTERVAL

Max check run interval

Default: `31536000`


### MAX_CHECK_QUEUE_TIME

Check queue timeout

Default: `120`


### MAX_CHECK_TIMEOUT

Max run timeout

Default: `300`


### MIN_CHECK_INTERVAL

Min check run interval

Default: `0`


### MIN_CHECK_TIMEOUT

Min run timeout

Default: `1`


### PREFER_CACHED_BROWSER


Whether the order of browser preference can be changed
to use a cached browser.

If a check supports Firefox and Chrome, firefox is used by default.

However, if an agent has a Chrome browser cached, enabling this option
will cause the run to switch to use Chrome, which will improve
run startup performance.

This may result in inconsistent browsers across runs, where a check is run
with one type of browser and a different on the next.


Default: `True`


### QUEUE_TASK_RESULT_RETENTION_MINS

Expiry time for celary task results from redis

Default: `1440`


### REDIS_HOST

Redis hostname/IP

Default: ``


### REDIS_INSTANCE

Redis instance ID

Default: ``


### REDIS_PASSWORD

Redis password

Default: ``


### REDIS_PORT

Redis port

Default: `0`


### REDIS_TYPE

Redis type - currently only supports 'redis'

Default: ``


### REDIS_USERNAME

Redis username

Default: ``


### RESULT_ARTIFACT_RETENTION_DAYS


Result artifact retention, containing logs and screenshots. This will replace any pre-existing S3 bucket policies.

Set to 0 if you are using a custom policy or do no want artifacts to be deleted.


Default: `365`


### RESULT_RETENTION_MINS


Expiry time of run results from database (in minutes).

Defaults to `RESULT_ARTIFACT_RETENTION_DAYS` (converted to minutes).

This should generally be configured to the same period of time as RESULT_ARTIFACT_RETENTION_DAYS.


Default: `525600`


### SCREENSHOT_ON_FAILURE_DEFAULT

Whether to always perform a screenshot after a run in browser tests

Default: `True`


### SENTRY_DSN

Sentry DSN to enable error reporting to sentry

Default: ``


### SENTRY_ENVIRONMENT

Sentry environment

Default: `prod`


### UI_RESULT_EXPIRE



Default: `604800`


### VICTORIAMETRICS_URL

Victoriametrics URL

Default: ``



CHANGELOG.md
# Changelog

# [4.18.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.17.1...v4.18.0) (2024-03-06)


### Features

* Move result metrics to victoriametrics ([d1f326f](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/d1f326feaf5433cf9edf29267a4319272d2295a5)), closes [#122](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/122)
* Update heatmap data API to use victoriametrics to obtain data ([2f46f2c](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/2f46f2ce0fe6f72ecbf4daea9b0cfcf0c0d63f18)), closes [#122](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/122)

## [4.17.1](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.17.0...v4.17.1) (2024-03-05)


### Bug Fixes

* Remove example sentry DNS ([6cb3755](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/6cb37559a6c77e143b3073286aa60a20ca35a445))

# [4.17.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.16.0...v4.17.0) (2024-03-04)


### Features

* Add heatmap timeframe for 15 minute intervals ([e695c3b](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/e695c3bd84b89d27757d3ff46f15407b68927b68)), closes [#120](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/120)

# [4.16.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.15.0...v4.16.0) (2024-03-03)


### Features

* Add check for verifying DNS response records ([5a768d9](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/5a768d91f4831024fb739aeebdf2c38c3ffe1505)), closes [#116](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/116)
* Add DNS step to perform DNS query ([56453bf](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/56453bf18ab975d46e829f9aff902f9eea6aa9a3)), closes [#116](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/116)
* Add support for checking CNAMEs in DNS resolution ([659f526](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/659f5263660f982264e2cd811899676189052ae4)), closes [#116](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/116)
* Add support for checking count and minimum count of DNS records ([06fc604](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/06fc604513f64ae6e8ef0b9d48b10c9e92da2c42)), closes [#116](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/116)

# [4.15.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.14.1...v4.15.0) (2024-03-02)


### Features

* Add Sentry integration ([397c4d5](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/397c4d52cfec9f0412949bc04a783ee8d076b76e)), closes [#119](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/119)

## [4.14.1](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.14.0...v4.14.1) (2024-02-21)


### Bug Fixes

* Add 'init' to agent container in docker compose. ([e4b3843](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/e4b384319397ed940b00f85aa71bf3dbff3909f0)), closes [#115](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/115)

# [4.14.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.13.0...v4.14.0) (2024-02-21)


### Features

* Add heatmap showing historical uptime percentages ([3d620bc](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/3d620bcb3ecc9f9734564d9b778662d4dc12ec26)), closes [#114](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/114)
* Store pass/fail in redis for each tiemframe to be shown on heatmap. Add draft API for returning data ([d1f26bd](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/d1f26bd58597bb0ac74d9d05fea359324d65caf1)), closes [#114](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/114)

# [4.13.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.12.1...v4.13.0) (2024-01-02)


### Features

* Add datetime picker to list view and to_date/from_date argument to run list API ([72e33cf](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/72e33cf5dc84880f1388fe4ddb5619ddd1884d27)), closes [#113](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/113)

## [4.12.1](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.12.0...v4.12.1) (2024-01-02)


### Bug Fixes

* Update background of graph in RunView to gray ([f743116](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/f7431162f8a0c51a0032239c7176eecd7f999789)), closes [#111](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/111)

# [4.12.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.11.7...v4.12.0) (2023-12-07)


### Bug Fixes

* Fix emoji used in slack message for recovery in example slack plugin ([7373707](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/7373707d961941ccc5a01a7390d9dcb707680120)), closes [#108](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/108)


### Features

* Add environment name, run timestamp to notification plugin calls. Add environment to slack notifications and link to failed runs. ([940405a](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/940405aa1d1b1d1f86ed4bb962ad2c7bb011344a)), closes [#107](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/107)
* Add notification hooks for timed out status of check. ([1a406bf](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/1a406bf7a091455e7a0d126d23bb9cacdbfee1c7)), closes [#109](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/issues/109)

## [4.11.7](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.11.6...v4.11.7) (2023-11-18)


### Bug Fixes

* **deps:** bump pyyaml from 6.0 to 6.0.1 ([ad2cdeb](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/ad2cdeb8fc98c4995de3ce7974505bd36ec821c0))

## [4.11.6](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/compare/v4.11.5...v4.11.6) (2023-10-01)


### Bug Fixes

* **deps:** bump @fontsource/roboto from 4.5.8 to 5.0.8 in /ui ([fa61fee](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/fa61feec333e8585c92bd36bb41696d1a19e9bf6))
* **deps:** bump @mui/material from 5.11.10 to 5.14.11 in /ui ([5a8bb69](https://gitlab.dockstudios.co.uk/pub/jmon/jmon/commit/5a8bb6916693e14ce886c4dba4f2c4a56f7cbfcc))

jmon/config.py

from enum import Enum
import os
from typing import Optional


class ChromeHeadlessMode(Enum):
    """Type of headless mode to use for chrome"""
    # Disable headless mode
    NONE = "None"
    # "New" chrome headless mode - this is a newer method, but appears to be slower
    NEW = "New"
    # Legacy chrome headless mode - this appears to be the fastest method
    LEGACY = "Legacy"


class Config:

    _INSTANCE: Optional['Config'] = None

    @property
    def DATABASE_TYPE(self) -> Optional[str]:
        """Database type - currently only 'postgresql' is supported"""
        return os.environ.get("DB_TYPE")

    @property
    def DATABASE_HOST(self) -> Optional[str]:
        """Database hostname/IP"""
        return os.environ.get("DB_HOST")

    @property
    def DATABASE_PORT(self) -> Optional[str]:
        """Database port"""
        return os.environ.get("DB_PORT")

    @property
    def DATABASE_USERNAME(self) -> Optional[str]:
        """Database username"""
        return os.environ.get("DB_USERNAME")

    @property
    def DATABASE_PASSWORD(self) -> Optional[str]:
        """Database password"""
        return os.environ.get("DB_PASSWORD")

    @property
    def DATABASE_NAME(self) -> Optional[str]:
        """Database DB name"""
        return os.environ.get("DB_NAME")

    @property
    def DEFAULT_CHECK_INTERVAL(self) -> int:
        """Default check run interval, if not specified by check (seconds)"""
        return int(os.environ.get('DEFAULT_CHECK_INTERVAL', '300'))

    @property
    def MAX_CHECK_INTERVAL(self) -> int:
        """Max check run interval"""
        return int(os.environ.get('MAX_CHECK_INTERVAL', '31536000'))

    @property
    def MIN_CHECK_INTERVAL(self) -> int:
        """Min check run interval"""
        return int(os.environ.get('MIN_CHECK_INTERVAL', '0'))

    @property
    def DEFAULT_CHECK_TIMEOUT(self) -> int:
        """Default run timeout"""
        return int(os.environ.get('DEFAULT_CHECK_TIMEOUT', '60'))

    @property
    def MAX_CHECK_TIMEOUT(self) -> int:
        """Max run timeout"""
        return int(os.environ.get('MAX_CHECK_TIMEOUT', '300'))

    @property
    def MIN_CHECK_TIMEOUT(self) -> int:
        """Min run timeout"""
        return int(os.environ.get('MIN_CHECK_TIMEOUT', '1'))

    @property
    def MAX_CHECK_QUEUE_TIME(self) -> int:
        """Check queue timeout"""
        return int(os.environ.get('MAX_CHECK_QUEUE_TIME', 120))

    # Default to 1 week for checks to expire in UI
    @property
    def UI_RESULT_EXPIRE(self):
        return int(os.environ.get("UI_RESULT_EXPIRE", "604800"))

    @property
    def API_KEY(self) -> Optional[str]:
        """
        API key

        Default unset, allowing unauthenticated access to APIs
        """
        return os.environ.get("API_KEY")

    @property
    def CHECK_CRITICAL_THRESHOLD(self) -> float:
        """Threshold for warning->critical check status in UI"""
        return float(os.environ.get("CHECK_CRITICAL_THRESHOLD", "99.0"))

    @property
    def CHECK_WARNING_THRESHOLD(self) -> float:
        """Threshold for okay->warning check status in UI"""
        return float(os.environ.get("CHECK_WARNING_THRESHOLD", "99.9"))

    @property
    def DATABASE_URL(self) -> str:
        """Return database url"""
        sqlalchemy_type = None
        if self.DATABASE_TYPE == "postgresql":
            sqlalchemy_type = "postgresql+psycopg2"

        if sqlalchemy_type is None:
            raise Exception("Unrecognised DB_TYPE")

        return f"{sqlalchemy_type}://{self.DATABASE_USERNAME}:{self.DATABASE_PASSWORD}@{self.DATABASE_HOST}:{self.DATABASE_PORT}/{self.DATABASE_NAME}"

    @property
    def BROKER_TYPE(self) -> str:
        """Configure type of broker. Currently only supports 'amqp'"""
        return os.environ.get('BROKER_TYPE', 'amqp')

    @property
    def BROKER_HOST(self) -> Optional[str]:
        """Broker hostname/ip"""
        return os.environ.get('BROKER_HOST')

    @property
    def BROKER_USERNAME(self) -> Optional[str]:
        """Broker username"""
        return os.environ.get('BROKER_USERNAME')

    @property
    def BROKER_PASSWORD(self) -> Optional[str]:
        """Broker password"""
        return os.environ.get('BROKER_PASSWORD')

    @property
    def BROKER_PORT(self) -> int:
        """Broker port"""
        return int(os.environ.get('BROKER_PORT', 0))

    @property
    def BROKER_INSTANCE(self) -> Optional[str]:
        """Broker vhost"""
        return os.environ.get('BROKER_INSTANCE')

    @property
    def REDIS_TYPE(self) -> Optional[str]:
        """Redis type - currently only supports 'redis'"""
        return os.environ.get('REDIS_TYPE')

    @property
    def REDIS_HOST(self) -> Optional[str]:
        """Redis hostname/IP"""
        return os.environ.get('REDIS_HOST')

    @property
    def REDIS_USERNAME(self) -> Optional[str]:
        """Redis username"""
        return os.environ.get('REDIS_USERNAME')

    @property
    def REDIS_PASSWORD(self) -> Optional[str]:
        """Redis password"""
        return os.environ.get('REDIS_PASSWORD')

    @property
    def REDIS_PORT(self) -> int:
        """Redis port"""
        return int(os.environ.get('REDIS_PORT', 0))

    @property
    def REDIS_INSTANCE(self) -> Optional[str]:
        """Redis instance ID"""
        return os.environ.get('REDIS_INSTANCE')

    @property
    def VICTORIAMETRICS_URL(self) -> Optional[str]:
        """Victoriametrics URL"""
        return os.environ.get("VICTORIAMETRICS_URL")

    @property
    def CACHE_BROWSER(self) -> bool:
        """
        Whether to cache browser between runs.

        This is experimental and might be unstable.

        Use on non-production environment
        """
        return os.environ.get("CACHE_BROWSER", "False") == "True"

    @property
    def PREFER_CACHED_BROWSER(self) -> bool:
        """
        Whether the order of browser preference can be changed
        to use a cached browser.

        If a check supports Firefox and Chrome, firefox is used by default.

        However, if an agent has a Chrome browser cached, enabling this option
        will cause the run to switch to use Chrome, which will improve
        run startup performance.

        This may result in inconsistent browsers across runs, where a check is run
        with one type of browser and a different on the next.
        """
        return os.environ.get("PREFER_CACHED_BROWSER", "True") == "True"

    @property
    def CHROME_HEADLESS_MODE(self) -> ChromeHeadlessMode:
        """Chrome headless mode. Options: None, New or Legacy"""
        return ChromeHeadlessMode(os.environ.get("CHROME_HEADLESS_MODE", "Legacy"))

    @property
    def FIREFOX_HEADLESS(self) -> bool:
        """Whether to run firefox in headless mode"""
        return os.environ.get("FIREFOX_HEADLESS", "True") == "True"

    @property
    def AWS_ENDPOINT(self) -> Optional[str]:
        """HTTPS url for AWS ENDPOINT for accessing S3. Set to minio API URL, if in use."""
        return os.environ.get('AWS_ENDPOINT')

    @property
    def AWS_BUCKET_NAME(self) -> Optional[str]:
        """Name of s3 bucket for storing artifacts"""
        return os.environ.get('AWS_BUCKET_NAME')

    @property
    def SCREENSHOT_ON_FAILURE_DEFAULT(self) -> bool:
        """Whether to always perform a screenshot after a run in browser tests"""
        return os.environ.get('SCREENSHOT_ON_FAILURE_DEFAULT', 'True').lower() == 'true'

    @property
    def QUEUE_TASK_RESULT_RETENTION_MINS(self) -> int:
        """Expiry time for celary task results from redis"""
        return int(os.environ.get("QUEUE_TASK_RESULT_RETENTION", "1440"))

    @property
    def RESULT_RETENTION_MINS(self) -> int:
        """
        Expiry time of run results from database (in minutes).

        Defaults to `RESULT_ARTIFACT_RETENTION_DAYS` (converted to minutes).

        This should generally be configured to the same period of time as RESULT_ARTIFACT_RETENTION_DAYS.
        """
        return int(os.environ.get("RESULT_RETENTION_MINS", self.RESULT_ARTIFACT_RETENTION_DAYS * 24 * 60))

    @property
    def RESULT_ARTIFACT_RETENTION_DAYS(self) -> int:
        """
        Result artifact retention, containing logs and screenshots. This will replace any pre-existing S3 bucket policies.

        Set to 0 if you are using a custom policy or do no want artifacts to be deleted.
        """
        return int(os.environ.get("RESULT_ARTIFACT_RETENTION_DAYS", "365"))

    @property
    def SENTRY_DSN(self) -> Optional[str]:
        """Sentry DSN to enable error reporting to sentry"""
        return os.environ.get("SENTRY_DSN")

    @property
    def SENTRY_ENVIRONMENT(self) -> str:
        """Sentry environment"""
        return os.environ.get("SENTRY_ENVIRONMENT", "prod")

    @classmethod
    def get(cls) -> 'Config':
        """Get instance of config"""
        if cls._INSTANCE is None:
            cls._INSTANCE = cls()
        return cls._INSTANCE
